### PR TITLE
Add a debug mode to browsertime with abitrary pausing.

### PR DESCRIPTION
**Pull request by gmierz**
This patch adds a --debug flag that can be used to enter debug mode. This mode enables a hotkey `ctrl+alt+g` which can be used to pause execution after the iteration is complete. It's also exposed in the test script so it can be used by custom test scripts as they may perform their own iterations.

Furthermore, when in --debug mode the Firefox browser will be detached from the terminal and open the devtools panel by default.

### CODE REVIEW

**Comment by soulgalore** [comment]
Thank you @gmierz this looks super cool! I'll will test it out later this week!

File: package.json [hunk 1]
```diff
@@ -27,6 +27,7 @@
     "lodash.pick": "4.4.0",
     "lodash.set": "4.3.2",
     "selenium-webdriver": "4.1.2",
+    "uiohook-napi": "^1.1.0",
```
Comment (soulgalore) [inline package.json:30]: I didn't get this to work on my Mac and it also seems to fail on the GitHub Action on Mac? I'm thinking two things:
* Maybe we could have it as an optional dependency (and follow the pattern we have for that)? But it should still work on all platforms, do you know if it should? I could also see in the package.json for that package that it seems to be NodeJS 16 and never.
* Lets sync how we should communicate between Browsertime and the browser. If we can't get keys to work, maybe we could use devtools and global variables in JS?

File: lib/firefox/webdriver/builder.js [hunk 2]
```diff
@@ -199,6 +199,11 @@ module.exports.configureBuilder = function (builder, baseDir, options) {
     }
   }
 
+  if (options.debug) {
```
Comment (soulgalore) [inline lib/firefox/webdriver/builder.js:202]: Think we need to check for not running on Android? Or do the devtools switch work on Android?

Comment (soulgalore) [reply lib/firefox/webdriver/builder.js:202]: Also I can add the same for Chrome.

**Review by soulgalore — CHANGES_REQUESTED** [review]
Check if you can get uiOhook to work, then I can try it some more!

File: package.json (hunk 1 above)
Comment (gmierz) [reply package.json:30]: Ah good catch! I was running in node-16 so maybe it pulled a version that only works for node-16.

**Comment by gmierz** [comment]
Thanks for the review @soulgalore :) I'll update the PR later next week

**Comment by soulgalore** [comment]
@gmierz I'm really found of this PR and I can help out getting things to work, but I need your help/guidance how we gonna do with the uiohook-nap library since it it fails to run on Mac (maybe others too)? One way could be do do this in multiple steps and first skip that lib and just communicate using devtools (or at least try it).

**Comment by gmierz** [comment]
@soulgalore regarding the new package, this was the only one I could find that worked for me. There are others but they are either broken, or only run on the client side of node (we can't use it within browsertime code as they require the DOM).

Can you elaborate on your idea with using devtools?

**Comment by soulgalore** [comment]
Ok! Let me try to make a POC, hope to have it ready Friday/Saturday.

**Comment by soulgalore** [comment]
@gmierz ok I tried now and maybe it's a different use case of what you try to do?

I've changed so that Firefox isn't detached from browsertime and then I added a debug command so that in your script where you want to debug it you add:
` await commands.debug(); ` (or pause or whatever) and then it sets browsertime.debug = true and loops until it's false. When my debug is finished, I set `browsertime.debug = false` and the script continous to the next pause/breakpoint in the script. 

To fix when we test one URL I guess we could add that command after the first URL is finished.

However it's not as fancy as stop/resume by clicking a button?

**Comment by soulgalore** [comment]
Another alternative could be for browser that supports it to load an extension when we run in debug mode that adds a continue button, so that you don't have to add JS directly in the console, I guess that would work?

**Comment by gmierz** [comment]
@soulgalore I think the devtools method you mentioned would work well, but we would need to keep Firefox detached from the terminal so that the browser doesn't close if the terminal process is killed. I think we can do that with your method as well though so there's no issues.

The method is very similar to what I had implemented and the only real difference is the lack of hotkeys. Using hotkeys, we have the same behaviour but it feels "fancier". We could also leave a debug loop after the iteration ends but before the browser stops and allow a user to continue by pressing enter in the console (this can be done easily with readline - it's global hotkeys that are an issue). This would result in almost exactly the same behaviour as we have with hotkeys but it would be slightly less dynamic which isn't a problem to me.

**Comment by soulgalore** [comment]
> but we would need to keep Firefox detached from the terminal so that the browser doesn't close if the terminal process is killed.

I didn't fully understand that part, how is that used? Like do you as a dev close the command line? I'm thinking now with the timeout fix you did the command line will just wait? 

Ok cool, if we can get rid of uiohook-napi so it can run on all platforms that's the most important for me. Do you want me to continue and can I push to your branch? You readline idea sounds good too!

**Comment by gmierz** [comment]
> > but we would need to keep Firefox detached from the terminal so that the browser doesn't close if the terminal process is killed.
> 
> I didn't fully understand that part, how is that used? Like do you as a dev close the command line? I'm thinking now with the timeout fix you did the command line will just wait?

Not the command line/terminal, but the browsertime node process. If someone were to exit that browsertime process, firefox would stay open when running in detached mode. Right now, if we kill browsertime, it also kills the browser session.

> 
> Ok cool, if we can get rid of uiohook-napi so it can run on all platforms that's the most important for me. Do you want me to continue and can I push to your branch? You readline idea sounds good too!

Yes please if you have some time! Otherwise, I'll get back to it later next week.

**Comment by soulgalore** [comment]
> Yes please if you have some time! Otherwise, I'll get back to it later next week

Ok, I'll try to do something this weekend then.

> Not the command line/terminal, but the browsertime node process. If someone were to exit that browsertime process, firefox would stay open when running in detached mode. Right now, if we kill browsertime, it also kills the browser session.

Yeah I got that but I don't understand why someone kill that process when you debug? :)

**Comment by gmierz** [comment]
Oh sorry! I'll find more information for you next week about this, but it's more for browser rather than web development. We do this in all of our existing test harnesses at Mozilla.

**Comment by soulgalore** [comment]
Pushed https://github.com/sitespeedio/browsertime/pull/1798 as the alternative

**Pull request closed without merging**